linera-service/tests/end_to_end_tests.rs

```diff
@@ -1581,145 +1657,395 @@ async fn test_wasm_end_to_end_amm(config: impl LineraNetConfig) -> Result<()> {
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::ZERO),
+        ])
+        .await;
 
-    app_fungible0_admin
+    app_amm1.swap(owner1, 0, Amount::from_tokens(50)).await?;
+    node_service_amm.process_inbox(&chain_amm).await?;
+    node_service1.process_inbox(&chain1).await?;
+
+    app_fungible0_amm
         .assert_balances([
             (owner0, Amount::ZERO),
-            (owner1, Amount::from_tokens(12)),
-            (owner_admin, Amount::ZERO),
-            (owner_amm, Amount::from_tokens(238)),
+            (owner1, Amount::ZERO),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::from_tokens(250)),
         ])
         .await;
-    app_fungible1_admin
+    app_fungible1_amm
+        .assert_balances([
+            (owner0, Amount::ZERO),
+            (owner1, Amount::ZERO),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::from_millis(102_400)),
+        ])
+        .await;
+
+    app_fungible0_0
+        .assert_balances([
+            (owner0, Amount::ZERO),
+            (owner1, Amount::ZERO),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::ZERO),
+        ])
+        .await;
+    app_fungible1_0
         .assert_balances([
             (owner0, Amount::from_tokens(50)),
-            (owner1, Amount::from_attos(104_957310924369747899)),
-            (owner_admin, Amount::ZERO),
-            (owner_amm, Amount::from_attos(95_042689075630252101)),
+            (owner1, Amount::ZERO),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::ZERO),
         ])
         .await;
 
-    app_amm0.swap(owner0, 1, Amount::from_tokens(50)).await;
-    node_service_admin.process_inbox(&chain_admin).await?;
+    app_fungible0_1
+        .assert_balances([
+            (owner0, Amount::ZERO),
+            (owner1, Amount::from_tokens(20)),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::ZERO),
+        ])
+        .await;
+    app_fungible1_1
+        .assert_balances([
+            (owner0, Amount::ZERO),
+            (owner1, Amount::from_millis(97_600)),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::ZERO),
+        ])
+        .await;
 
-    app_fungible0_admin
+    app_amm0.swap(owner0, 1, Amount::from_tokens(50)).await?;
+    node_service_amm.process_inbox(&chain_amm).await?;
+    node_service0.process_inbox(&chain0).await?;
+
+    app_fungible0_amm
         .assert_balances([
-            (owner0, Amount::from_attos(82_044810916287757646)),
-            (owner1, Amount::from_tokens(12)),
-            (owner_admin, Amount::ZERO),
-            (owner_amm, Amount::from_attos(155_955189083712242354)),
+            (owner0, Amount::ZERO),
+            (owner1, Amount::ZERO),
+            (owner_amm_chain, Amount::ZERO),
+            (owner_amm_app, Amount::from_attos(167_979002624671916011)),
```

Comment:
No, I agree with that, that's a good idea!
I'm saying it would be even better if all the resulting numbers in the example were whole, or at least had a low number of decimal digits. So it would be nicer if we could work out numbers for how much liquidity to provide, how many tokens to swap, and how much to withdraw, in such a way that the result is always an integer or something. Not important, though. One just has to use a calculator if one really wants to double-check these numbers. :smile: